src/python/WMCore/MicroService/Unified/MSRuleCleaner.py

```diff
@@ -50,27 +225,296 @@ def execute(self, reqStatus):
+        in the pipeline.
+        :param  wflow:   A MSRuleCleaner workflow representation
+        :param  pName:   The name of the functional pipeline
+        :return wflow:
+        """
+        # NOTE: The current functional pipeline MUST always be appended at the
+        #       end of the 'PlineMarkers' list
+
+        # First check if we can safely get rid of the default:
+        if len(wflow['PlineMarkers']) == 1 \
+           and wflow['PlineMarkers'][-1] is None \
+           and pName is not None:
+            wflow['PlineMarkers'].pop()
+
+        # Then push our current value into the markers list:
+        wflow['PlineMarkers'].append(pName)
+        return wflow
+
+    def _checkClean(self, wflow):
+        """
+        An auxiliary function used to only check the temporary cleanup status.
+        It basically takes the pipelines registered in 'PlineMarkers' that have
+        already worked on the workflow as a mask and applies this mask over
+        the set of flags in the 'CleanupStatus' field and then reduces the
+        result to a single bool value
+        """
+        # NOTE: This is one of the few functions taking a workflow as an argument
+        #       but returning a bool, since it is an auxiliary function and is not
+        #       supposed to be called as a standalone function in a pipeline.
+        # NOTE: `all([]) == True`, ergo all the 'rejected' && 'aborted-completed' workflows
+        #       are also counted as properly cleaned and can trigger archival later
+        cleanStatus = all([wflow['CleanupStatus'][key]
+                           for key in wflow['PlineMarkers']
+                           if key in wflow['CleanupStatus'].keys()])
+        return cleanStatus
+
+    def setClean(self, wflow):
+        """
+        A function to set the 'IsClean' flag based on the status from all the
+        pipelines which have worked on the workflow (and have put their markers
+        in the 'PlineMarkers' list)
+        :param  wflow:      A MSRuleCleaner workflow representation
+        :return wflow:
+        """
+        wflow['IsClean'] = self._checkClean(wflow)
+        return wflow
+
+    def archive(self, wflow):
+        """
+        Move the workflow to the proper archived status after checking
+        the full cleanup status
+        :param  wflow:      A MSRuleCleaner workflow representation
+        :param  archStatus: Target status to transition after archival
+        :return wflow:
+        """
+        # NOTE: check allowed status transitions with:
+        #       https://github.com/dmwm/WMCore/blob/5961d2229b1e548e58259c06af154f33bce36c68/src/python/WMCore/ReqMgr/DataStructs/RequestStatus.py#L171
+        if not wflow['IsClean']:
+            msg = "Not properly cleaned workflow: %s" % wflow['RequestName']
+            raise MSRuleCleanerArchival(msg)
+
+        # Check the available status transitions before we decide the final status
```

Comment:
I'd add something like:
```
if check_allowed_transition(wflow['RequestStatus'], wflow['TargetStatus']) is False:
  BIG exception
```
otherwise, do the request transition (and make a log record)

Reply:
Keep that though - to be worked in the next GH issue.